internal/blocktx/processor.go

```diff
@@ -188,45 +187,56 @@ func (p *Processor) StartBlockProcessing() {
 			case <-p.ctx.Done():
 				return
 			case blockMsg := <-p.blockProcessCh:
+				blockhash := blockMsg.Header.BlockHash()
+				p.logger.Info("received block", slog.String("hash", blockhash.String()))
+
 				err := p.processBlock(blockMsg)
+
 				if err != nil {
-					blockhash := blockMsg.Header.BlockHash()
-					_, errDel := p.store.DelBlockProcessing(p.ctx, &blockhash, p.hostname)
-					if errDel != nil {
-						p.logger.Error("failed to delete block processing", slog.String("hash", blockhash.String()), slog.String("err", errDel.Error()))
-					}
+					p.logger.Error("block processing failed", slog.String("hash", blockhash.String()), slog.String("err", err.Error()))
+					p.unlockBlock(p.ctx, &blockhash)
 				}
 			}
 		}
 	}()
 }
 
-func (p *Processor) StartFillGaps(peers []p2p.PeerI) {
+// unlock block for future processing
+func (p *Processor) unlockBlock(ctx context.Context, hash *chainhash.Hash) {
+	// use closures for retries
+	unlockFn := func() error {
+		_, err := p.store.DelBlockProcessing(ctx, hash, p.hostname)
+		return err
+	}
+
+	if unlockErr := withRetry(ctx, unlockFn, 5); unlockErr != nil {
+		p.logger.ErrorContext(ctx, "failed to delete block processing", slog.String("hash", hash.String()), slog.String("err", unlockErr.Error()))
+	}
+}
+
+func (p *Processor) startBlockProcessedGuard(ctx context.Context, hash *chainhash.Hash) {
 	p.waitGroup.Add(1)
 
-	ticker := time.NewTicker(p.fillGapsInterval)
-	go func() {
+	go func(ctx context.Context, hash *chainhash.Hash) {
 		defer p.waitGroup.Done()
-		peerIndex := 0
-		for {
-			select {
-			case <-p.ctx.Done():
-				return
-			case <-ticker.C:
-				if peerIndex >= len(peers) {
-					peerIndex = 0
-				}
 
-				err := p.fillGaps(peers[peerIndex])
-				if err != nil {
-					p.logger.Error("failed to fill gaps", slog.String("error", err.Error()))
-				}
+		// respect processor shutdown while waiting
+		select {
+		case <-p.ctx.Done():
+			return // we can do nothing here
```

Comment:
Shouldn't we also run `p.unlockBlock(ctx, hash)` before exiting the function in case that the go routine stops due to a cancelled context? 🤔

Reply:
I made some changes to the logic and added more descriptive comments.

In summary, this goroutine will stop (via its context) when block processing is complete or when the processor shuts down, so we don't want to unlock the blocks here. I moved the unlocking of all blocks in progress to the Shutdown function.